Move rate limiting to swarm to fix the unprotected buffer

ii-cruz opened the pull request:
## What's in this pull request?
Moving rate limiting to swarm to fix the unprotected buffer of the network when handling a request.
Rate limit refactoring

#### This fixes #2531.

## Pull request checklist
- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

ii-cruz committed:
Moving rate limiting to swarm.

nibhar commented on network-libp2p/src/network_types.rs:
```diff
@@ -248,8 +250,14 @@ pub(crate) struct TaskState {
+        (
+            mpsc::Sender<(Bytes, InboundRequestId, PeerId)>,
+            u32,
+            Duration,
```
I would prefer these be encapsulated into a named struct. It is used in a couple more places than just here. Not tagging all of them but just here.

nibhar commented on network-libp2p/src/rate_limiting.rs:
```diff
@@ -142,10 +147,13 @@ impl RateLimit {
 
-pub(crate) fn is_under_the_rate_limits<Req: RequestCommon>(
-    peer_request_limits: Arc<Mutex<HashMap<PeerId, HashMap<u16, RateLimit>>>>,
+pub(crate) fn is_under_the_rate_limits(
```
Personally i would always go for `exceeds_rate_limit()`. Obviously that would inverse the return. however you already inverse the return from `increment_and_is_allowed` and then when calling this you invert again, so it should be fine,.

That being said doing all of this caller side would remove some of the clutter. The function has a bunch of parameters which are just used to init or retrieve the `RateLimit`. I would probably inline this making the function as a whole obsolete. Specifically as it is only used in a single spot.

nibhar reviewed (approved):
Generally I would like this to be more encapsulated. It looks correct to me and if necessary could go in like this, but I think it would be worth our while.

ii-cruz committed:
Implement feedback

nibhar commented on network-libp2p/src/network_types.rs:
```diff
@@ -248,8 +250,13 @@ pub(crate) struct TaskState {
     pub(crate) response_channels:
         HashMap<InboundRequestId, ResponseChannel<Option<OutgoingResponse>>>,
     /// Senders for replying to requests per `RequestType` for request-response
```
This should probably reflect the update made.

nibhar committed:
Remove Mutex from RateLimit as it has &mut self everywhere.

Co-authored-by: ii-cruz <ii.pintocruz@gmail.com>

nibhar commented on network-libp2p/src/rate_limiting.rs:
```diff
@@ -141,114 +163,120 @@ impl RateLimit {
+
+        // Ensures that the request is allowed based on the set limits and updates the counter.
+        // Returns early if not allowed.
+        if !requests_limit.increment_and_is_allowed(1) {
+            info!(
+                %request_id,
+                %peer_id,
+                %request_type,
+                "Rate limit was exceeded!",
+            );
+            log::debug!(
+                "[{:?}][{:?}] {:?} Exceeded max requests rate {:?} requests per {:?} seconds",
+                request_id,
+                peer_id,
+                request_type,
+                request_rate_limit_data.max_requests,
+                request_rate_limit_data.time_window,
+            );
+            return true;
+        }
+
+        false
```
I suggest keeping the debug log only. it is more informativ. Also I'd suggest to use tracing notation. Also if you log this callerside this can be replaced by 
```suggestion
        !requests_limit.increment_and_is_allowed(1)
```

nibhar reviewed (commented):
Smaller notes but nothing major. I did not mention this in the review itself but add some rustdoc to functions and fields please.

hrxi committed:
Only accept control transactions over the control transaction topic

Do this by using a wrapper that only accepts control transactions.

Fixes #2614.

hrxi commented on network-interface/src/request/mod.rs:
```diff
@@ -143,7 +143,7 @@ pub trait RequestCommon:
     /// Serializes a request.
     /// A serialized request is composed of:
-    /// - A varint for the Type ID of the request
+    /// - A variant for the Type ID of the request
```
```suggestion
    /// - A variable sized integer for the Type ID of the request
```

`varint` as in "variable sized integer".

ii-cruz replied:
Thanks this was autocorrect mix-fixes

hrxi commented on network-libp2p/src/swarm.rs:
```diff
@@ -702,53 +700,64 @@ fn handle_event(
+                                    let sender_data = state
+                                        .receive_requests
+                                        .get(&type_id)
+                                        .filter(|(sender, ..)| !sender.is_closed());
```
This changes behavior. Previously, closed channels were removed from the map.

Is the behavior change intentional?

ii-cruz replied:
They are being removed later yes. I discussed it with JD. The code becomes more readable like this and the removal still happens later.

hrxi commented on network-libp2p/src/swarm.rs:
```diff
@@ -1204,6 +1222,29 @@ pub(crate) fn verify_record(record: &Record) -> Option<DhtRecord> {
+            "Could not send response {:?}", response
+        );
+    }
+}
```
I feel like this function isn't carrying its weight. What was the reason for extracting the code into a function? Logging a message on error?

ii-cruz replied:
it's repeated in two places close by and simplified a bit the code there because there are so many If, else closes and nested scopes. Can change it back if you prefer, either way is fine imo

ii-cruz replied:
I removed the function and the response and respective error handling are done in the original places

hrxi commented on network-interface/src/request/mod.rs:
```diff
@@ -26,17 +26,17 @@ impl fmt::Display for RequestType {
-    pub const fn message(type_id: u16) -> RequestType {
-        RequestType::new(type_id, false)
+    pub const fn message(type_id: u16) -> Self {
+        Self::new(type_id, false)
```
This seems unrelated to the PR. Was it done unintentionally using automated refactoring?

ii-cruz replied:
I saw this issue on some other structs related to the rate-limiting and spotted it here too so I changed it as well

hrxi reviewed (commented):
Found a couple of potential issues.

dependabot[bot] committed:
build(deps): bump docker/build-push-action from 5.4.0 to 6.0.0

Bumps [docker/build-push-action](https://github.com/docker/build-push-action) from 5.4.0 to 6.0.0.
- [Release notes](https://github.com/docker/build-push-action/releases)
- [Commits](https://github.com/docker/build-push-action/compare/ca052bb54ab0790a636c9b5f226502c73d547a25...c382f710d39a5bb4e430307530a720f50c2d3318)

---
updated-dependencies:
- dependency-name: docker/build-push-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

nibhar committed:
continue polling Tendermint when old state update is received.

hrxi committed:
Add some tests around control transactions

hrxi commented:
Rebasing to merge.

viquezclaudio committed:
Fix clippy warning

nibhar committed:
Tendermint: Choose better future contribution to verify than just random

Don't verify contributions whose contributors are a subset of the already verified contributors.

dependabot[bot] committed:
build(deps): bump derive_more in the minor-and-patch-updates group

Bumps the minor-and-patch-updates group with 1 update: [derive_more](https://github.com/JelteF/derive_more).


Updates `derive_more` from 0.99.17 to 0.99.18
- [Release notes](https://github.com/JelteF/derive_more/releases)
- [Changelog](https://github.com/JelteF/derive_more/blob/v0.99.18/CHANGELOG.md)
- [Commits](https://github.com/JelteF/derive_more/compare/v0.99.17...v0.99.18)

---
updated-dependencies:
- dependency-name: derive_more
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: minor-and-patch-updates
...

Signed-off-by: dependabot[bot] <support@github.com>
